radix-engine/src/blueprints/resource/fungible/fungible_bucket.rs

```diff
@@ -10,41 +10,50 @@ use radix_engine_interface::blueprints::resource::*;
 
 pub struct FungibleBucket;
 
-impl FungibleBucket {
-    pub fn liquid_amount<Y>(api: &mut Y) -> Result<Decimal, RuntimeError>
+pub struct FungibleBucketBlueprint;
+
+impl FungibleBucketBlueprint {
+    fn get_divisibility<Y>(api: &mut Y) -> Result<u8, RuntimeError>
     where
-        Y: KernelNodeApi + ClientApi<RuntimeError>,
+        Y: ClientApi<RuntimeError>,
     {
-        let handle = api.actor_open_field(
-            OBJECT_HANDLE_SELF,
-            FungibleBucketField::Liquid.into(),
+        let divisibility_handle = api.actor_open_field(
+            OBJECT_HANDLE_OUTER_OBJECT,
+            FungibleResourceManagerField::Divisibility.into(),
             LockFlags::read_only(),
         )?;
-        let substate_ref: LiquidFungibleResource = api.field_lock_read_typed(handle)?;
-        let amount = substate_ref.amount();
-        api.field_lock_release(handle)?;
-        Ok(amount)
+        let divisibility: u8 = api.field_lock_read_typed(divisibility_handle)?;
+        api.field_lock_release(divisibility_handle)?;
+        Ok(divisibility)
     }
 
-    pub fn locked_amount<Y>(api: &mut Y) -> Result<Decimal, RuntimeError>
+    pub fn take<Y>(amount: Decimal, api: &mut Y) -> Result<Bucket, RuntimeError>
     where
         Y: KernelNodeApi + ClientApi<RuntimeError>,
     {
-        let handle = api.actor_open_field(
-            OBJECT_HANDLE_SELF,
-            FungibleBucketField::Locked.into(),
-            LockFlags::read_only(),
-        )?;
-        let substate_ref: LockedFungibleResource = api.field_lock_read_typed(handle)?;
-        let amount = substate_ref.amount();
-        api.field_lock_release(handle)?;
-        Ok(amount)
+        Self::take_advanced(amount, WithdrawStrategy::Exact, api)
     }
 
-    pub fn take<Y>(amount: Decimal, api: &mut Y) -> Result<LiquidFungibleResource, RuntimeError>
+    pub fn take_advanced<Y>(
+        amount: Decimal,
+        withdraw_strategy: WithdrawStrategy,
+        api: &mut Y,
+    ) -> Result<Bucket, RuntimeError>
     where
         Y: KernelNodeApi + ClientApi<RuntimeError>,
     {
+        // Apply withdraw strategy
+        let divisibility = Self::get_divisibility(api)?;
+        let amount = amount.for_withdrawal(divisibility, withdraw_strategy);
```

Comment:
I imagine given how often this code path will be hit with `WithdrawStrategy::Exact` (eg from the `take` call), it's worth us skipping the divisibility read in this case?

Probably simplest is to inline `for_withdrawal` for performance, and only doing the divisibility read where it's required.

Reply:
Although actually I guess we still need to do the `check_fungible_amount` call anyway with the divisibility, so no biggy!

Reply:
Yeah, we need that for amount checking anyway.